File: frontend/lib/schemas/v2/poolBulk.ts
/**
 * プール一括投入 (W-2) zod スキーマ.
 *
 * BE 契約 (設計書 `docs/plans/pool-bulk-insert-design.md` §4 を正典とする):
 *   POST /api/v1/schedule/v2/pool-bulk-simulate (read-only)
 *   POST /api/v1/schedule/v2/pool-bulk-apply    (1 TX 一括適用)
 *   RBAC: admin / manager (BE 側で 403 担保).
 *
 * BE simulate schema (`backend/app/schemas/v2/pool_bulk.py`, コミット 1cb702d) と 1:1。
 * apply の Request/Response は設計書 §4 を正典とする (BE 並行実装中)。
 *
 * 寛容パース規約 (improvementSuggestion の kind 方式踏襲):
 *   - placements / partial / unplaced は要素単位で safeParse し、パース不能な要素
 *     (= 未知 enum 値等) を静かに除外する。これによりセクション全滅を防ぐ。
 *   - warnings 系は `.catch([])`、KPI 欠落フィールドは `.default()` で drift 吸収。
 *   - start_time は "HH:MM:SS" (pool-overview と同形式) をそのまま文字列保持。
 */
import { z } from 'zod';

import { v2WeekdayBeforeAfterSchema } from './autoScheduleV2';

/** pool-bulk API 上限 (BE POOL_BULK_MAX_PATIENTS と同値。超過は 422). */
export const POOL_BULK_MAX_PATIENTS = 50;

// ---------------------------------------------------------------------------
// 寛容パースヘルパー (未知要素をセクション全滅させず静かに除外)
// ---------------------------------------------------------------------------

/**
 * 配列を要素単位で safeParse し、パース不能な要素を除外する寛容配列スキーマ。
 * improvementSuggestion.ts の tolerantSuggestionsArraySchema と同じ設計。
 */
function tolerantArray<T extends z.ZodTypeAny>(schema: T) {
  return z.preprocess((val) => {
    if (!Array.isArray(val)) return val;
    return val.filter((el) => schema.safeParse(el).success);
  }, z.array(schema));
}

// ---------------------------------------------------------------------------
// Response sub-models
// ---------------------------------------------------------------------------

/** 投入成功した 1 枠 (患者 × 曜日 × コース). seq は投入順の通し番号. */
export const poolBulkPlacementSchema = z.object({
  seq: z.number().int(),
  patient_id: z.string(),
  patient_name: z.string(),
  weekday: z.number().int().min(0).max(6),
  course_code: z.string(),
  office_id: z.string(),
  /** "HH:MM:SS" (pool-overview と同形式). 表示側で秒を落とす. */
  start_time: z.string(),
  service_minutes: z.number().int(),
  /** 挿入の厳密限界コスト (分/週). 個別提案と同一の物差し. 未評価は null. */
  delta_minutes: z.number().nullable().optional().default(null),
  /** propose-slots と同一語彙の警告コード (staff_absent 等). */
  warnings: z.array(z.string()).catch([]),
});
export type PoolBulkPlacement = z.infer<typeof poolBulkPlacementSchema>;

/** 部分投入 (不足曜日の一部だけ入った) 患者 1 名の要約. */
export const poolBulkPartialSchema = z.object({
  patient_id: z.string(),
  patient_name: z.string(),
  placed_days: z.number().int(),
  missing_days: z.number().int(),
  /** 投入できなかった希望曜日 (str) → 理由コード (excluded_summary と同語彙). */
  unplaced_reasons: z.record(z.string(), z.string()).catch({}),
});
export type PoolBulkPartial = z.infer<typeof poolBulkPartialSchema>;

/** 投入不能 (1 枠も入らなかった) 患者 1 名の要約. */
export const poolBulkUnplacedSchema = z.object({
  patient_id: z.string(),
  patient_name: z.string(),
  /** 理由コード (capacity_full/travel_shortage/lunch_window/no_gap/course_closed/no_coordinates). */
  reason: z.string(),
});
export type PoolBulkUnplaced = z.infer<typeof poolBulkUnplacedSchema>;

/** プレビュー用サマリ KPI (投入件数 + 週全体の移動時間/距離 before→after). */
export const poolBulkKpiSchema = z.object({
  placed_patients: z.number().int().default(0),
  placed_slots: z.number().int().default(0),
  travel_minutes_before: z.number().default(0),
  travel_minutes_after: z.number().default(0),
  travel_km_before: z.number().default(0),
  travel_km_after: z.number().default(0),
});
export type PoolBulkKpi = z.infer<typeof poolBulkKpiSchema>;

// ---------------------------------------------------------------------------
// simulate: Request / Response
// ---------------------------------------------------------------------------

export const poolBulkSimulateRequestSchema = z.object({
  iso_year: z.number().int(),
  iso_week: z.number().int().min(1).max(53),
  /** 対象拠点 (単一必須). state_token が拠点単位の指紋のため. */
  office_id: z.string(),
  /** 保留プール患者 id 列 (FE が抽出して送る. pool-overview と同方式). 最大 50 件. */
  patient_ids: z.array(z.string()).max(POOL_BULK_MAX_PATIENTS),
  /** 投入順序. v1 は hybrid のみ (D-1). */
  ordering: z.literal('hybrid').default('hybrid'),
});
export type PoolBulkSimulateRequest = z.input<typeof poolBulkSimulateRequestSchema>;

export const poolBulkSimulateResponseSchema = z.object({
  placements: tolerantArray(poolBulkPlacementSchema).default([]),
  partial: tolerantArray(poolBulkPartialSchema).default([]),
  unplaced: tolerantArray(poolBulkUnplacedSchema).default([]),
  /** ProposalWeekCalendar / BeforeAfterWeekPanel 互換の before/after 週ビュー. */
  week_before_after: tolerantArray(v2WeekdayBeforeAfterSchema).default([]),
  kpi: poolBulkKpiSchema.default({}),
  /** 楽観ロック用指紋 (apply で再計算し不一致なら 409). */
  state_token: z.string(),
});
export type PoolBulkSimulateResponse = z.infer<typeof poolBulkSimulateResponseSchema>;

// ---------------------------------------------------------------------------
// apply: Request / Response
// ---------------------------------------------------------------------------

/**
 * POST /v2/pool-bulk-apply リクエスト (設計書 §4).
 *   - placements: simulate の placements をそのまま送る (プレフィックス選択は v1 なし = 全件).
 *   - change_scope は持たせない = 仕様として pattern_and_week (A) 固定 (D-2).
 *   - state_token 不一致は BE が 409 を返す (FE は再計算導線を出す).
 */
export const poolBulkApplyRequestSchema = z.object({
  iso_year: z.number().int(),
  iso_week: z.number().int().min(1).max(53),
  office_id: z.string(),
  placements: z.array(poolBulkPlacementSchema),
  state_token: z.string(),
});
export type PoolBulkApplyRequest = z.input<typeof poolBulkApplyRequestSchema>;

/** POST /v2/pool-bulk-apply レスポンス (設計書 §4). */
export const poolBulkApplyResponseSchema = z.object({
  applied_patients: z.number().int().default(0),
  applied_slots: z.number().int().default(0),
  warnings: z.array(z.string()).catch([]),
});
export type PoolBulkApplyResponse = z.infer<typeof poolBulkApplyResponseSchema>;

File: backend/app/schemas/v2/pool_bulk.py
"""プール一括投入 (W-1: simulate) Pydantic v2 schemas.

エンドポイント:
  - POST /api/v1/schedule/v2/pool-bulk-simulate (admin/manager, read-only)

保留プールの患者列を **逐次シミュレーション** で 1 人ずつ決定論的に積み、投入先
(placements) / 部分投入 (partial) / 投入不能 (unplaced) / 前後 KPI / before-after 週ビュー /
state_token を返す read-only API. 物差し (実現可能性 + delta 分/週) は個別提案
(propose-slots) と完全に同一 (``compute_all_proposed_slots`` を再利用).

設計書 ``docs/plans/pool-bulk-insert-design.md`` §3 (エンジン) / §4 (API 契約).

実装方針:
  - simulate は read-only (DB 書込みなし). apply (W-2) と完全分離.
  - サイズガード: ``len(patient_ids) > POOL_BULK_MAX_PATIENTS`` は 422
    (pool-overview の ``POOL_OVERVIEW_MAX_PATIENTS`` と同値).
"""

from __future__ import annotations

import uuid
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field, field_validator

from app.schemas.v2.auto_schedule_v2 import V2WeekdayBeforeAfter

# 単一ソースはエンジン. スキーマは再エクスポートのみ (二重定義排除).
from app.services.scheduling.pool_bulk_inserter import POOL_BULK_MAX_PATIENTS

# ---------------------------------------------------------------------------
# Request
# ---------------------------------------------------------------------------


class PoolBulkSimulateRequest(BaseModel):
    """``POST /v2/pool-bulk-simulate`` リクエスト (対象週 × 拠点 × プール患者 id 列)."""

    model_config = ConfigDict(extra="forbid")

    iso_year: int = Field(..., ge=2020, le=2100)
    iso_week: int = Field(..., ge=1, le=53)
    # 対象拠点 (単一必須). 一括投入は単一拠点内で完結する (設計書 §7).
    # state_token が拠点単位の指紋のため office_id は必須.
    office_id: uuid.UUID = Field(..., description="対象拠点 (単一)")
    # 保留プール患者 id 列 (FE が抽出して送る. pool-overview と同方式).
    patient_ids: list[uuid.UUID] = Field(default_factory=list)
    # 投入順序. v1 は hybrid のみ (D-1). 将来拡張用に enum.
    ordering: Literal["hybrid"] = "hybrid"

    @field_validator("patient_ids")
    @classmethod
    def _guard_size(cls, v: list[uuid.UUID]) -> list[uuid.UUID]:
        if len(v) > POOL_BULK_MAX_PATIENTS:
            raise ValueError(
                f"patient_ids は最大 {POOL_BULK_MAX_PATIENTS} 件までです (received: {len(v)} 件)"
            )
        return v


# ---------------------------------------------------------------------------
# Response
# ---------------------------------------------------------------------------


class PoolBulkPlacement(BaseModel):
    """投入成功した 1 枠 (患者 × 曜日 × コース). seq は投入順の通し番号."""

    model_config = ConfigDict(extra="forbid")

    seq: int = Field(..., ge=1, description="投入順の通し番号 (1..)")
    patient_id: uuid.UUID
    patient_name: str
    weekday: int = Field(..., ge=0, le=6, description="0=Mon..6=Sun")
    course_code: str
    office_id: uuid.UUID
    start_time: str = Field(..., description="HH:MM:SS (pool-overview と同形式)")
    service_minutes: int = Field(..., ge=1, le=480)
    delta_minutes: float | None = Field(
        default=None, description="挿入の厳密限界コスト (分/週). 個別提案と同一の物差し"
    )
    warnings: list[str] = Field(
        default_factory=list, description="propose-slots と同一語彙の警告コード (staff_absent 等)"
    )


class PoolBulkPartial(BaseModel):
    """部分投入 (不足曜日の一部だけ入った) 患者 1 名の要約."""

    model_config = ConfigDict(extra="forbid")

    patient_id: uuid.UUID
    patient_name: str
    placed_days: int = Field(..., ge=1, description="投入できた枠数")
    missing_days: int = Field(..., ge=1, description="投入できなかった不足枠数")
    unplaced_reasons: dict[str, str] = Field(
        default_factory=dict,
        description="投入できなかった希望曜日 (str) → 理由コード (excluded_summary と同語彙)",
    )
    # A 案 (2026-07-04 PO 承認): 超過は一括に組み込まず個別の相談フロー (方式b) へ橋渡しする.
    overcapacity_available_count: int = Field(
        default=0,
        ge=0,
        description="定員起因 (capacity_full) の未充足曜日に「定員 +1 なら入る候補」がある件数 "
        "(>=1 なら方式b の相談で採用可能)",
    )


class PoolBulkUnplaced(BaseModel):
    """投入不能 (1 枠も入らなかった) 患者 1 名の要約."""

    model_config = ConfigDict(extra="forbid")

    patient_id: uuid.UUID
    patient_name: str
    reason: str = Field(
        ...,
        description="理由コード (capacity_full/travel_shortage/lunch_window/no_gap/"
        "course_closed/no_coordinates/no_primary_office/office_mismatch/two_staff_pending)",
    )
    # A 案 (2026-07-04 PO 承認): capacity_full のとき「定員 +1 なら入る候補」件数を載せ、個別フローへ橋渡し.
    overcapacity_available_count: int = Field(
        default=0,
        ge=0,
        description="capacity_full で投入不能な患者に「定員 +1 なら入る候補」がある件数 "
        "(>=1 なら方式b の相談で採用可能)",
    )


class PoolBulkKpi(BaseModel):
    """プレビュー用サマリ KPI (投入件数 + 週全体の移動時間/距離 before→after)."""

    model_config = ConfigDict(extra="forbid")

    placed_patients: int = Field(..., ge=0, description="1 枠以上投入できた患者数")
    placed_slots: int = Field(..., ge=0, description="投入枠の合計数")
    travel_minutes_before: int = Field(..., ge=0)
    travel_minutes_after: int = Field(..., ge=0)
    travel_km_before: float = Field(..., ge=0.0)
    travel_km_after: float = Field(..., ge=0.0)


class PoolBulkSimulateResponse(BaseModel):
    """``POST /v2/pool-bulk-simulate`` レスポンス (read-only プレビュー)."""

    model_config = ConfigDict(extra="forbid")

    placements: list[PoolBulkPlacement] = Field(default_factory=list)
    partial: list[PoolBulkPartial] = Field(default_factory=list)
    unplaced: list[PoolBulkUnplaced] = Field(default_factory=list)
    # ProposalWeekCalendar / FullOptimizeDialog 互換の before/after 週ビュー.
    week_before_after: list[V2WeekdayBeforeAfter] = Field(default_factory=list)
    kpi: PoolBulkKpi
    # 楽観ロック用指紋 (apply=W-2 で再計算し不一致なら 409).
    state_token: str


# ---------------------------------------------------------------------------
# Apply (W-2)
# ---------------------------------------------------------------------------


class PoolBulkApplyRequest(BaseModel):
    """``POST /v2/pool-bulk-apply`` リクエスト (設計書 §4).

    simulate の ``placements`` をそのまま送る. **change_scope フィールドは持たせない**:
    一括投入は 1 件ずつ聞けないため反映先は pattern_and_week 固定 (D-2). state_token を
    サーバで再計算し、不一致なら 409 (simulate 後にスケジュールが変わったら必ずやり直し).
    """

    model_config = ConfigDict(extra="forbid")

    iso_year: int = Field(..., ge=2020, le=2100)
    iso_week: int = Field(..., ge=1, le=53)
    office_id: uuid.UUID = Field(..., description="対象拠点 (単一・state_token の単位)")
    # simulate の placements をそのまま (プレフィックス選択は v1 なし = 全件).
    placements: list[PoolBulkPlacement] = Field(
        default_factory=list,
        max_length=350,
        description="投入対象の配置リスト (上限 350 = 最大 50 患者 × 7 曜日).",
    )
    # 楽観ロック用指紋 (simulate が返したもの). サーバ再計算と不一致なら 409.
    state_token: str
    # NG スタッフ / 性別制限の確認フロー (docs/plans/patient-ng-staff-design.md §7-2)。
    # simulate は NG 枠を提案しないが、simulate 後に NG が登録されると古い placements の
    # apply が素通りするため、**患者ループの前に全 placements をまとめて事前検査** する。
    acknowledge_constraint_warnings: bool = Field(
        default=False,
        description="NGスタッフ/性別制限の警告を確認済みとして続行する (422 の再送用)。",
    )


class PoolBulkApplyResponse(BaseModel):
    """``POST /v2/pool-bulk-apply`` レスポンス (適用サマリ)."""

    model_config = ConfigDict(extra="forbid")

    applied_patients: int = Field(..., ge=0, description="固定枠を登録できた患者数")
    applied_slots: int = Field(..., ge=0, description="登録した固定枠 (placements) の合計数")
    warnings: list[str] = Field(
        default_factory=list,
        description="pfv_validator V3-V5 (衝突/昼休み/容量) の warning 文言 (ブロックしない)",
    )


__all__ = [
    "POOL_BULK_MAX_PATIENTS",
    "PoolBulkApplyRequest",
    "PoolBulkApplyResponse",
    "PoolBulkKpi",
    "PoolBulkPartial",
    "PoolBulkPlacement",
    "PoolBulkSimulateRequest",
    "PoolBulkSimulateResponse",
    "PoolBulkUnplaced",
]
